### PR TITLE
RestartIO: Handle INT vectors

### PR DESCRIPTION
**Pull request by akva2**
This adds writing of INT restart vectors using visitors. I've included #3278 for simplicity. If you want, we can handle that one separately first.

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
jenkins build this please

**Comment by bska** [comment]
> I've included #3278 for simplicity. If you want, we can handle that one separately first.

Right.  Yes, I think I'd prefer to review that part in isolation before getting on to the new functionality here.

**Comment by akva2** [comment]
jenkins build this please

**Review by bska — COMMENTED** [review]
This looks good to me.  I'd like a to know why we're moving away from forwarding references for the callback functions though.

File: opm/output/data/Cells.hpp [hunk 1]
```diff
@@ -124,6 +124,12 @@ namespace Opm { namespace data {
             return std::get<std::vector<T>>(data_);
         }
 
+        template<class Visitor>
+        void visit(Visitor visit) const
+        {
+            std::visit(visit, data_);
+        }
```
[the remark this answers is not in the document]
Comment (akva2) [reply opm/output/data/Cells.hpp]: sure.

File: src/opm/output/eclipse/RestartIO.cpp [hunk 2]
```diff
@@ -614,33 +623,43 @@ namespace {
...
                               const std::vector<std::string>& vectors,
-                              const bool                      write_double,
-                              OutputVector&&                  writeVector)
+                              OutputVector                    writeVectorF,
+                              OutputVectorInt                 writeVectorI)
     {
         for (const auto& vector : vectors) {
-            writeVector(vector, value.solution.data<double>(vector), write_double);
+            value.solution.at(vector).visit(VisitorOverloadSet{
+                MonoThrowHandler<std::logic_error>{"Attempt to access monostate field."},
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/opm/output/eclipse/RestartIO.cpp]: sure.

File: src/opm/output/eclipse/RestartIO.cpp [hunk 3]
```diff
@@ -614,33 +623,43 @@ namespace {
...
         }
     }
 
-    template <class OutputVector>
+    template <class OutputVector, class OutputVectorInt>
     void writeRegularSolutionVectors(const RestartValue& value,
-                                     const bool          write_double,
-                                     OutputVector&&      writeVector)
+                                     OutputVector        writeVectorF,
+                                     OutputVectorInt     writeVectorI)
```
[the remark this answers is not in the document]
Comment (akva2) [reply src/opm/output/eclipse/RestartIO.cpp]: no, got a bit eager when removing forwarding in the main writeSolutionVectors. remedied.

**Comment by akva2** [comment]
jenkins build this please

File: src/opm/output/eclipse/RestartIO.cpp [hunk 4]
```diff
@@ -614,33 +623,45 @@ namespace {
...
     void writeSolutionVectors(const RestartValue&             value,
                               const std::vector<std::string>& vectors,
-                              const bool                      write_double,
-                              OutputVector&&                  writeVector)
+                              OutputVector                    writeVectorF,
+                              OutputVectorInt                 writeVectorI)
     {
         for (const auto& vector : vectors) {
-            writeVector(vector, value.solution.data<double>(vector), write_double);
+            value.solution.at(vector).visit(std::move(VisitorOverloadSet{
```
Comment (bska) [inline src/opm/output/eclipse/RestartIO.cpp:633]: I don't think you need `std::move()` here.  `VisitorOverloadSet {...}` will be an x-value in the call to `CellData::visit()` which template argument deduction will perfectly forward without the explicit `move()`.

**Review by bska — COMMENTED** [review]
Thanks a lot for the updates.  I think there's one potential improvement left, but I'd be prepared to merge this version as well.

File: src/opm/output/eclipse/RestartIO.cpp (hunk 4 above)
Comment (akva2) [reply src/opm/output/eclipse/RestartIO.cpp]: indeed. i had it removed, forgot to recommit.

**Comment by bska** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks a lot for the updates.  This looks good to me now and I'll merge into master.

**Pull request merged**